docgen/src/guide/Refining_hits.md

```diff
@@ -0,0 +1,591 @@
+  state = {
+    refinements: {
+      keywords: [],
+      "owner.name": [],
+    },
+  };
+
+  onRefine = ({ attributeName, value }) =>
+    this.setState(({ refinements }) => ({
+      refinements: {
+        ...refinements,
+        [attributeName]: [...value],
+      },
+    }));
+
+  refine = ({ attributeName, value }) =>
+    this.setState(({ refinements }) => {
+      const index = refinements[attributeName].indexOf(value);
+
+      // if it's not yet refined, add it to the refinements
+      if (index === -1) {
+        const oldValue = refinements[attributeName];
+        return {
+          refinements: {
+            ...refinements,
+            [attributeName]: [...oldValue, value],
+          },
+        };
+      }
+
+      // if it's already refined, remove it from the refinements
+      const newValue = refinements[attributeName].slice();
+      newValue.splice(index, 1);
+      return {
+        refinements: {
+          ...refinements,
+          [attributeName]: newValue,
+        },
+      };
+    });
+
+  render() {
+    const { refinements } = this.state;
+    return (
+      <InstantSearch
+        appId="OFCNCOG2CU"
+        apiKey="f54e21fa3a2a0160595bb058179bfb1e"
+        indexName="npm-search"
+      >
+        <Configure
+          attributesToRetrieve={[
+            "name",
+            "description",
+            "keywords",
+            "owner.name",
+          ]}
+        />
+        <SearchBox />
+        <CurrentRefinements />
+        <div style={{ display: "flex", border: "1px solid black" }}>
+          <RefinementList attributeName="keywords" withSearchBox />
+          <RefinementList attributeName="owner.name" withSearchBox />
+        </div>
+        <VirtualRefinementList
+          attributeName="keywords"
+          defaultRefinement={refinements.keywords}
+          onRefine={this.onRefine}
+        />
+        <VirtualRefinementList
+          attributeName="owner.name"
+          defaultRefinement={refinements["owner.name"]}
+          onRefine={this.onRefine}
+        />
+        <Hits onRefine={this.refine} />
+      </InstantSearch>
+    );
+  }
+}
+
+export default App;
+```
+
+We notice that in `CurrentRefinements`, each list show up twice. This is because the default current refinements widget will render each widget's refinements by default, and we have both a `RefinementList` and a `VirtualRefinementList` for the same attribute. As you might see in the console, this causes "duplicate key" errors by React. To prevent those, we can use `transformData` on `CurrentRefinements`. Each `item` of the current refinements has an `index` and `id`. These two properties explain which index the widget works on (see [multi index](guide/Multi_index.html)) and which the identifier of the widget has (this can be `'query'`, `attributeName`, `'page'` and some others). We can deduplicate the current refinements using those two properties like this:
+
+```jsx
+// ...
+const deleteDuplicates = items =>
```

Comment:
this is wrong, should be: 

```js
export const deleteDuplicates = (items) =>
  items
    .map(item => {
      // make a temporary key for deduplication
      item.__dedupe = `${item.index}.${item.id}`;
      return item;
    })
    .filter(
      (obj, pos, arr) =>
        arr.map(item => item.__dedupe).indexOf(obj.__dedupe) === pos
    )
    .map(item => {
      // delete it afterwards
      delete item.__dedupe;
      return item;
    });
```

Reply:
```js
export const deleteDuplicates = (items) =>
  items
    .filter(
      (obj, pos, arr) =>
        arr.map(item => `${item.index}.${item.id}`).indexOf(`${obj.index}.${obj.id}`) === pos
    )
```

two less iterations (still O(n!))

Reply:
Since the array is pretty small (less than 50 elements most of the time) I think the more understandable is the best 🙂